File: app/security/crypto_manager.py
"""
crypto_manager.py

AES-256-GCM application-layer payload encryption, compatible with the
companion Java implementation in java_client/CryptoManager.java.

Wire format (all bytes, then base64url-encoded as a single string):

    [ 12-byte nonce ][ ciphertext ][ 16-byte GCM auth tag ]

This matches the standard output layout of Java's
Cipher.getInstance("AES/GCM/NoPadding") (ciphertext + 16-byte tag
concatenated) prefixed with the nonce we generate ourselves, and it
matches PyCryptodome / `cryptography`'s AESGCM, which also appends the
tag to the ciphertext.

Security notes:
- Uses a fresh random 12-byte nonce for every encryption call, as
  required for AES-GCM.
- Uses the standard 16-byte (128-bit) authentication tag.
- Never re-uses a nonce with the same key.
- The key itself is NOT generated or stored here - it's provided by
  app.config.get_settings().PAYLOAD_ENCRYPTION_KEY (see config.py and
  the README section on key management / session credentials).
"""
import base64
import hashlib
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM
from cryptography.exceptions import InvalidTag

NONCE_SIZE = 12  # bytes, standard for GCM
TAG_SIZE = 16    # bytes, standard GCM tag length


class DecryptionError(Exception):
    """Raised when a payload cannot be decrypted or its integrity check fails."""
    pass


def _derive_key(raw_key: str) -> bytes:
    """
    Normalizes an arbitrary-length secret string into exactly 32 bytes
    (AES-256) using SHA-256. This lets the operator supply any
    reasonably strong secret (e.g. a token_urlsafe(32) string) via the
    PAYLOAD_ENCRYPTION_KEY env var without worrying about exact byte
    length. The Java side must perform the identical derivation
    (SHA-256 of the UTF-8 key string) to stay compatible.
    """
    return hashlib.sha256(raw_key.encode("utf-8")).digest()


def encrypt_data(data: str, key: str) -> str:
    """
    Encrypts a UTF-8 string and returns a base64url (no padding) string
    containing nonce || ciphertext || tag.
    """
    aes_key = _derive_key(key)
    nonce = os.urandom(NONCE_SIZE)
    aesgcm = AESGCM(aes_key)
    ciphertext_and_tag = aesgcm.encrypt(nonce, data.encode("utf-8"), None)
    blob = nonce + ciphertext_and_tag
    return base64.urlsafe_b64encode(blob).decode("ascii").rstrip("=")


def decrypt_data(encrypted_data: str, key: str) -> str:
    """
    Reverses encrypt_data(). Raises DecryptionError if the payload is
    malformed or the authentication tag does not verify (i.e. the data
    was tampered with, corrupted, or encrypted with a different key).
    """
    aes_key = _derive_key(key)
    try:
        padded = encrypted_data + "=" * (-len(encrypted_data) % 4)
        blob = base64.urlsafe_b64decode(padded.encode("ascii"))
    except Exception as exc:
        raise DecryptionError(f"Malformed base64 payload: {exc}") from exc

    if len(blob) < NONCE_SIZE + TAG_SIZE:
        raise DecryptionError("Payload too short to contain nonce + tag")

    nonce = blob[:NONCE_SIZE]
    ciphertext_and_tag = blob[NONCE_SIZE:]

    aesgcm = AESGCM(aes_key)
    try:
        plaintext = aesgcm.decrypt(nonce, ciphertext_and_tag, None)
    except InvalidTag as exc:
        raise DecryptionError("Authentication tag verification failed") from exc
    except Exception as exc:
        raise DecryptionError(f"Decryption failed: {exc}") from exc

    return plaintext.decode("utf-8")

File: java_client/CryptoManager.java
package com.hexprotocol.client;

import javax.crypto.Cipher;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.Arrays;
import java.util.Base64;

/**
 * CryptoManager.java
 *
 * AES-256-GCM application-layer payload encryption, compatible with the
 * companion Python implementation in app/security/crypto_manager.py.
 *
 * Wire format (base64url, no padding, single string):
 *   [ 12-byte nonce ][ ciphertext ][ 16-byte GCM auth tag ]
 *
 * IMPORTANT - key management:
 * Do NOT hardcode PAYLOAD_ENCRYPTION_KEY as a string constant in this
 * class or anywhere in the shipped APK. A key baked into the APK can be
 * extracted by decompiling the app, which defeats the purpose of
 * encrypting the payload. Instead:
 *
 *   1. Preferred: have the server issue a short-lived, per-device
 *      session key after an initial authenticated handshake (e.g. over
 *      TLS, tied to a signed device attestation), store it only in
 *      memory / Android Keystore, and rotate it periodically.
 *   2. At minimum: pull the key from a secure remote config service at
 *      runtime rather than embedding it in source, and rely on HTTPS/TLS
 *      as the primary transport protection (this app-layer encryption is
 *      defense-in-depth, not a replacement for TLS).
 *
 * This class only implements the cipher operations - it deliberately
 * takes the key as a parameter rather than owning a hardcoded secret.
 */
public final class CryptoManager {

    private static final int NONCE_SIZE = 12;   // bytes
    private static final int TAG_SIZE_BITS = 128; // 16-byte GCM tag
    private static final String CIPHER_ALGO = "AES/GCM/NoPadding";
    private static final String KEY_ALGO = "AES";

    private final byte[] aesKey; // 32 bytes, derived from the raw secret

    /**
     * @param rawKey the shared secret string (e.g. the session key issued
     *               by the server). Normalized to 32 bytes via SHA-256 to
     *               match the Python side's key derivation exactly.
     */
    public CryptoManager(String rawKey) {
        this.aesKey = deriveKey(rawKey);
    }

    private static byte[] deriveKey(String rawKey) {
        try {
            MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
            return sha256.digest(rawKey.getBytes(StandardCharsets.UTF_8));
        } catch (Exception e) {
            throw new RuntimeException("Failed to derive AES key", e);
        }
    }

    /**
     * Encrypts a plaintext string and returns a base64url (no padding)
     * string containing nonce || ciphertext || tag.
     */
    public String encrypt(String plainData) {
        try {
            byte[] nonce = new byte[NONCE_SIZE];
            new SecureRandom().nextBytes(nonce);

            Cipher cipher = Cipher.getInstance(CIPHER_ALGO);
            SecretKeySpec keySpec = new SecretKeySpec(aesKey, KEY_ALGO);
            GCMParameterSpec spec = new GCMParameterSpec(TAG_SIZE_BITS, nonce);
            cipher.init(Cipher.ENCRYPT_MODE, keySpec, spec);

            byte[] cipherTextAndTag = cipher.doFinal(plainData.getBytes(StandardCharsets.UTF_8));

            ByteBuffer buffer = ByteBuffer.allocate(nonce.length + cipherTextAndTag.length);
            buffer.put(nonce);
            buffer.put(cipherTextAndTag);

            return base64UrlEncodeNoPadding(buffer.array());
        } catch (Exception e) {
            throw new RuntimeException("Encryption failed", e);
        }
    }

    /**
     * Reverses encrypt(). Throws RuntimeException (wrapping the
     * underlying cause) if the payload is malformed or the
     * authentication tag fails to verify - treat any exception here as
     * "reject this response/request", never fall back to using
     * unauthenticated data.
     */
    public String decrypt(String encryptedData) {
        try {
            byte[] blob = base64UrlDecode(encryptedData);
            if (blob.length < NONCE_SIZE + 16) {
                throw new IllegalArgumentException("Payload too short to contain nonce + tag");
            }

            byte[] nonce = Arrays.copyOfRange(blob, 0, NONCE_SIZE);
            byte[] cipherTextAndTag = Arrays.copyOfRange(blob, NONCE_SIZE, blob.length);

            Cipher cipher = Cipher.getInstance(CIPHER_ALGO);
            SecretKeySpec keySpec = new SecretKeySpec(aesKey, KEY_ALGO);
            GCMParameterSpec spec = new GCMParameterSpec(TAG_SIZE_BITS, nonce);
            cipher.init(Cipher.DECRYPT_MODE, keySpec, spec);

            byte[] plainBytes = cipher.doFinal(cipherTextAndTag);
            return new String(plainBytes, StandardCharsets.UTF_8);
        } catch (Exception e) {
            throw new RuntimeException("Decryption failed - payload may be corrupted, tampered with, or encrypted with a different key", e);
        }
    }

    private static String base64UrlEncodeNoPadding(byte[] data) {
        return Base64.getUrlEncoder().withoutPadding().encodeToString(data);
    }

    private static byte[] base64UrlDecode(String data) {
        return Base64.getUrlDecoder().decode(data);
    }

    // --------------------------------------------------------------
    // Example usage (remove or adapt in your real application code):
    //
    //   CryptoManager crypto = new CryptoManager(sessionKey);
    //   String payload = "{\"license_key\":\"HEX-XXXX-XXXX-XXXX\",\"installation_id\":\"...\",\"app_version\":\"1.0.0\"}";
    //   String encrypted = crypto.encrypt(payload);
    //   // send `encrypted` as the request body to POST /api/v1/auth/verify
    //   // ... receive `encryptedResponse` from the server ...
    //   String decrypted = crypto.decrypt(encryptedResponse);
    //   // parse `decrypted` as JSON and handle SUCCESS / INVALID_KEY / etc.
    // --------------------------------------------------------------
}
